#164530145 Implement recipe fetching functionality

devskope opened the pull request:
**What does this PR do?**
Adds Recipe fetching functionality

**Description of Task to be completed?**
- Modify slug generation strategy
- Implement functionality to
  - Get all recipes
  - Get recipes by slugs

**How should this be manually tested?**
- Checkout `ft-fetch-recipes-164530145`
- Run `npm i`
- Create a new user account by sending a valid post request to `/api/users`
- Create new recipes by sending valid post requests to `/api/recipes`
- Test fetching all recipes by sending a get request to `/api/recipes`
- Test fetching a recipe by its `slug` by sending a get request to `/api/recipes/${slug}`

**Any background context you want to provide?**
Currently, slugs are generated before saving recipe instances. After saving, the id of the newly saved instance is appended to its slug then the instance gets updated once more.

This PR moves all the logic associated with slug generation during recipe creation into the recipe model's `beforeSave` hook (a lifecycle event that is triggered before the creation of new objects in the database)  and appends a random 8 character hex string to the recipe's slug.


**What are the relevant pivotal tracker stories?**
[164530145](https://www.pivotaltracker.com/story/show/164530145)

**Screenshots (if appropriate)**

Commit:
feat(Recipes): implement recipe fetching:

- get all recipes
- make recipe slugs unique
- get recipes by slugs

devskope commented on server/controllers/recipe-controller.js:
```diff
@@ -54,6 +68,52 @@ class RecipeController {
+      .then(recipe =>
+        recipe
+          ? successResponse(res, { recipe })
+          : errorResponse(res, 'recipe Not found', 404)
```
🙄

Commit:
feat(Recipes): implement recipe fetching
- fix tests
- get recipes by slugs

LordUche reviewed (commented):
Looks good 👍

LordUche commented on test/recipes.spec.js:
```diff
@@ -2,48 +2,63 @@ import chai, { expect } from 'chai';
+            .that.has.lengthOf(0);
+          expect(message)
+            .to.be.a('string')
+            .that.eqls('no recipes have been created');
```
This test is too explicit. The actual value of the string could change at any time, so there is no need to test for it.

devskope replied:
I disagree.  I think it is sane to strictly verify outputs when possible. 
I don't think the value of this string will ever change if it does, then this will be updated.

andela-pessien commented on test/recipes.spec.js:
```diff
@@ -2,48 +2,63 @@ import chai, { expect } from 'chai';
+
 describe('Recipes', () => {
+  describe('Fetching before create', () => {
+    it('Get all should return empty array and message when no recipes exist', done => {
```
API tests are written by endpoint. I'm not sure why they're separated from their endpoint here

devskope replied:
I'll regroup these tests by endpoint

devskope replied:
@andela-pessien  I don't understand the Hound issues......  I have no linter errors locally and my indentations/whitespaces conform to the style guide

andela-pessien commented on server/utils/helpers.js:
```diff
@@ -105,3 +105,35 @@ export const generateVerificationLink = token =>
+ * @returns {string} Slugified string
+ */
+/* eslint-disable no-useless-escape */
+export const slugifyTitle = string =>
```
Why does this show up as a new change?

devskope replied:
It was moved from a separate file into helpers and documented

andela-pessien commented on server/models/recipe.js:
```diff
@@ -18,6 +18,15 @@ module.exports = (sequelize, DataTypes) => {
+      { slug: `${recipe.slug}-${recipe.id}` },
+      { returning: true }
+    );
+    Object.assign(recipe, updatedRecipe);
```
I feel like there should be a better way to do this,

devskope replied:
I think it can be done beforeCreate, and without the id..... 
A helper function could be whipped up to tack a random sequence(`length === 10`?)  of alphanumeric ([a-z 0-9]) characters onto the end of each slug beforeCreate.

andela-pessien commented on package.json:
```diff
@@ -33,15 +33,14 @@
     "swagger-ui-express": "4.0.2",
-    "underscore": "1.9.1"
+    "underscore": "1.9.1",
+    "webpack-merge": "4.2.1"
```
Why webpack-merge here? You have to admit it looks weird in a backend project 🙂

devskope replied:
Overkill maybe, basically just needed to meld objects together to make for composable sequelize query filters/options.

 &nbsp; Basically `{ ...commonOptions, ...extra }`

 &nbsp;&nbsp; &nbsp;              _Oh . ☝️  could actually replace that,_    given trivial shallow merges(which is basically what has been done at this [point](https://github.com/andela/vale-ah-backend/pull/19#discussion_r266396033))

andela-pessien commented on server/controllers/recipe-controller.js:
```diff
@@ -1,14 +1,28 @@
-const { Recipe } = db;
+const { Recipe, User } = db;
+
+const defaultRecipeDbFilter = {
```
Nice :)

Commit:
feat(recipes): implement recipe fetching

 - rebase with develop
 - fix tests

andela-pessien reviewed (changes requested):
Also please address the Hound issues.

Commit:
feat(recipes): implement recipe fetching

- reorganize test files
- discard webpack-mege
- modify slug generation strategy

[finishes #164530145]

andela-pessien commented on server/models/recipe.js:
```diff
@@ -20,24 +20,23 @@ module.exports = (sequelize, DataTypes) => {
-    Object.assign(recipe, updatedRecipe);
+
+  Recipe.beforeCreate(recipe => {
+    recipe.slug = `${slugifyTitle(recipe.title)}-${randomHex().substr(0, 8)}`;
```
Why is this change

1) introduced
2) in this PR
3) in this location?

devskope replied:
1.  The change was introduced to avoid hitting the database twice (first to create a recipe, then to update the recipe's slug).

2. Unique slugs were first added to this codebase by this PR. 

3. I also reasoned that unique slug generation should be handled all in one place right before saving the recipe into the database

andela-pessien replied:
Unique slugs were actually introduced in a different PR, as evidenced by the fact that they are already present in the develop branch. Here you are proposing a change to code that has already been merged. 

In essence be sure to pay attention to the commits you've just rebased your branch onto, not just accept or reject the changed from them. It should be clear, either from the PR or from code comments, why_modifications_ to existing code are being made, and preferably someone shouldn't have to ask :)

devskope replied:
Okay, got that. 

The PR description has been updated to reflect the reasons behind the modifications made.
